Add support for IndexIDMap with Cagra fp16

jinsolp opened the pull request:
This PR does 2 things
- Enable support fot `IndexIDMap` with Cagra fp16 (original support introduced in https://github.com/facebookresearch/faiss/pull/4188)
    - Added tests in `test_cagra.py`
- Reflecting feedback about python API from https://github.com/facebookresearch/faiss/pull/4384#discussion_r2167013214

mdouze reviewed (changes requested):
For this change to work without user visible degradation, we should implement implicit conversion fp16->fp32 in C++ for add/search/train. There are a few fp16 routines for CPU here: https://github.com/facebookresearch/faiss/blob/main/faiss/utils/fp16.h

jinsolp commented:
@mdouze Thanks for your feedbak! I have made a few changes : )
- Renamed all extended APIs with the `Ex` postfix because of cpp compile warnings related to not overriding overloaded functions.
- Added the `numeric_type` tag support for `torch_utils.py`

mnorris11 commented:
Sorry for delayed responses. At least to me, it does not seem like a problem to annotate this 1 line on all the indexes that require it, if the API stays as `add()`, etc. It looks cleaner and keeps with the current pattern.

jinsolp commented:
@mnorris11 Now the diff is really big, but we are 1) keeping the previous API (without the `Ex`) and 2) compiling without warnings.

swig wasn't able to process macros properly. So what I've done is manually override functions if the child struct is already overriding its original counterpart. (i.e. if the child class is already overriding `add()` I also overrode `add(void*, numeric_type)`. 
All overrode functions are just forwarded to its base struct's function to keep the logic consistent as before.
